### ISSUE TITLE
rclone mount deadlocks after cancelling upload

### ISSUE THREAD

**Issue by neocxf** [rclone/rclone#7081]
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/rclone-hungs-afther-long-usage/38453

#### What is the problem you are having with rclone?

rclone hangs when I do `ls`

#### What is your rclone version (output from `rclone version`)

`v1.54.0` with commit id `8a429d12cfb181bf3a28ce4e794b3eada8bf7e0e` 

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

ubuntu 18.04

#### Which cloud storage system are you using? (e.g. Google Drive)

aws s3

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

rclone mount 

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

![image](https://github.com/rclone/rclone/assets/12878595/79090b1e-117b-4e1e-9a6c-a14ff9aa1b90)

[rclone dump code](https://drive.google.com/file/d/15_gVc7T8wZm4cwHP1zDme5205n3Vbc74/view?usp=drive_link)

**Comment by neocxf** [comment]
@ncw

**Comment by ncw** [comment]
You wrote in the forum

> After reevaluating the code, I found a potential deadlock issue when using the default writeback mode.
> 
> As the pic show, the dir's mutex and cache's mutex may affect each other occasionally and cause a deadlock issue.
> Everytime such case happens, the whole mount system become unresponsible and have to be restarted to make the deadlock issue disappear.
> 
> I have examined the latest rclone code, I think such issue still remains

![无标题-2023-05-25-1559](https://github.com/rclone/rclone/assets/536803/23c9fabb-c608-452e-a6d2-bff140fe3583)

That is very helpful - thank you.

I think there are two issues here

1. `Item.RemoveNotInUse` is attempting to remove a file which is being uploaded and therefore in use and this should never happen.
2. The `Cache` lock is held for far too long while `Cache.purgeOld` is running.

I think fixing 1. first would be best. Somehow `item.info.Dirty` became false before the file was uploaded.

I think to fix this I'm going to need a way of triggering the problem.

Looking at the trace the problem will trigger when `--vfs-cache-mode writes` or `--vfs-cache-mode full` is in effect.

This problem will be more frequent if  `--vfs-cache-poll-interval` is decreased so say `1s` or `1ms`

      --vfs-cache-poll-interval Duration       Interval to poll the cache for stale objects (default 1m0s)


What do you think happened to the `meta-data.json` file? Did it get written and re-written maybe?

I could ideally do with a complete log with `-vv` of this happening. I don't care how long the log is!

If you could paste

1. The command line you used with rclone
2. A redacted (with secrets removed) copy of your rclone config file
3. A log with `-vv` of the problem happening
4. If you still have it the complete log of the original problem you posted (so I can trace what happened to `meta-data.json`)

Thank you.

**Comment by neocxf** [comment]
1. The command line I use is 
```shell
  rclone mount xxxxx:"${BUCKET_NAME}" /xxxxx-mnt --config /rclone/conf/rclone.conf \
    --cache-dir /tmp/rclone/.cache \
    --dir-cache-time 2s \
    --poll-interval 1s \
    --vfs-cache-mode writes \
    --vfs-cache-poll-interval 1s \
    --vfs-cache-max-age 60s \
    --log-file /var/log/rclone/rclone-mount.log \
    --allow-non-empty --no-modtime
```

2. The rclone conf file I use is 
```conf
[xxxxx]
type = s3
provider = AWS
access_key_id = ${ACCESS_KEY}
secret_access_key = ${SECRET_KEY}
region = ${BUCKET_REGION}
endpoint = ${BUCKET_ENDPOINT}
```
3.  The log file is: 
    [rclone-mount.log](https://drive.google.com/file/d/1gvMag5ftTQ-6hTdrLuhoa4ALFwVnCCQ-/view?usp=drive_link)
As the log shows, the last log remain as `meta-data.json: vfs cache: cancelling upload`

**Comment by ncw** [comment]
Here is my analysis of the log.

```
2023/05/09 19:10:06 DEBUG : /: Lookup: name="bills"
2023/05/09 19:10:06 DEBUG : : Re-reading directory (2.493826526s old)
```

The middle go routine is now holding the Dir lock

This function never returns as we don't see a `>Lookup` in the logs afterwards.

```
2023/05/09 19:10:08 DEBUG : meta-data.json: vfs cache: starting upload
2023/05/09 19:10:08 DEBUG : meta-data.json: md5 = b2fb13b2a88a6bdf4079fbf8bcc4d75f OK
2023/05/09 19:10:08 INFO  : meta-data.json: Copied (replaced existing)
2023/05/09 19:10:08 DEBUG : meta-data.json: vfs cache: fingerprint now "32,2023-05-09 19:10:03.94336583 +0000 UTC,b2fb13b2a88a6bdf4079fbf8bcc4d75f"
2023/05/09 19:10:08 DEBUG : meta-data.json: vfs cache: writeback object to VFS layer
```

The right hand go routine is now waiting for the Dir lock


```
2023/05/09 19:10:09 DEBUG : vfs cache RemoveNotInUse (maxAge=60000000000, emptyOnly=false): item meta-data.json not removed, freed 0 bytes
2023/05/09 19:10:09 INFO  : vfs cache: cleaned: objects 5 (was 5) in use 0, to upload 0, uploading 1, total size 137.589Ki (was 137.589Ki)
```

These two lines repeated once per second until

```
2023/05/09 19:11:03 DEBUG : vfs cache RemoveNotInUse (maxAge=60000000000, emptyOnly=false): item meta-data.json not removed, freed 0 bytes
2023/05/09 19:11:03 INFO  : vfs cache: cleaned: objects 3 (was 3) in use 0, to upload 0, uploading 1, total size 4.855Ki (was 4.855Ki)
2023/05/09 19:11:04 DEBUG : meta-data.json: vfs cache: cancelling writeback (uploading true) 0xc000a859d0 item 138
2023/05/09 19:11:04 DEBUG : meta-data.json: vfs cache: cancelling upload
```

The left hand process is now deadlocked

I think this deadlock was fixed with this commit 591fc3609aea47441fbd4bc01235291efea9bb4e which makes sure the object is not marked clean until after it was written back to the VFS layer and hence is not eligible to be purged so the left hand go routine will never run.

This was released in v1.61.0 so if you upgrade to that version or newer the problem should be fixed.

**Comment by neocxf** [comment]
Hi, @ncw 

I have tested with the `v1.62.2`. The problem seems still remained. 

Here is the version I use:

![image](https://github.com/rclone/rclone/assets/12878595/24f58217-6f0c-44ac-8c2c-519a4ebdd377)

Here is the rclone conf:

```conf
[xxxxx]
type = s3
provider = AWS
access_key_id = ${ACCESS_KEY}
secret_access_key = ${SECRET_KEY}
region = ${BUCKET_REGION}
endpoint = ${BUCKET_ENDPOINT}
```

Here is the rclone mount command:

```shell
  rclone mount xxxx:"${BUCKET_NAME}" /yyyy --config /rclone/conf/rclone.conf \
    --cache-dir /tmp/rclone/.cache \
    --dir-cache-time 2s \
    --poll-interval 1s \
    --vfs-cache-mode writes \
    --vfs-cache-poll-interval 1s \
    --vfs-cache-max-age 60s \
    --log-file /var/log/rclone/yyyy.log \
    --allow-non-empty --no-modtime
```

Here is the rclone log file:

[rclone.log](https://drive.google.com/file/d/1DTYEXZG-jb4eWwKnpVMC35Hmgs_6ZlU5/view?usp=drive_link)

**Comment by ncw** [comment]
> Here is the rclone log file:
> 
> [rclone.log](https://drive.google.com/file/d/1DTYEXZG-jb4eWwKnpVMC35Hmgs_6ZlU5/view?usp=drive_link)

That file appears to be corrupted?

```
$ ls -l rclone.log.tgz 
-rw-rw-r-- 1 ncw ncw 16862716 Jul 22 14:22 rclone.log.tgz
$ md5sum rclone.log.tgz 
fab6171b85fa1294254383e145ad71a9  rclone.log.tgz
$ tar xf rclone.log.tgz 

gzip: stdin: invalid compressed data--crc error

gzip: stdin: invalid compressed data--length error
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

**Comment by neocxf** [comment]
The file did got corrupted. I have updated the link. Here is the corret log file:

[rclone.log](https://drive.google.com/file/d/1DhRRMWGOxfDUs2MmTywtjuQhFb94E49A/view?usp=drive_link)

The checksum of the file is:

```shell
~/tmp » md5sum rclone.log.tar.gz                                                                                                       
599f53a6dabfd3d52355838ca7a46572  rclone.log.tar.gz
```

**Comment by ncw** [comment]
None of these logs are with v1.62.2

```
$ grep rclone: mount-billing.log 
2023/05/09 04:31:03 DEBUG : rclone: Version "v1.57.0" starting with parameters ["rclone" "mount" "fastone:fastone-billing-mywmicro-fcce-north1-1" "/fastone-mnt" "--config" "/rclone/conf/rclone.conf" "--cache-dir" "/tmp/rclone/.cache" "--dir-cache-time" "2s" "--poll-interval" "1s" "--vfs-cache-mode" "writes" "--vfs-cache-poll-interval" "1s" "--vfs-cache-max-age" "60s" "--log-file" "/var/log/rclone/mount-billing.log" "--allow-non-empty" "-vvvv"]
2023/05/16 08:33:35 DEBUG : rclone: Version "v1.57.0" starting with parameters ["rclone" "mount" "fastone:fastone-billing-mywmicro-fcce-north1-1" "/fastone-mnt" "--config" "/rclone/conf/rclone.conf" "--cache-dir" "/tmp/rclone/.cache" "--dir-cache-time" "2s" "--poll-interval" "1s" "--vfs-cache-mode" "writes" "--vfs-cache-poll-interval" "1s" "--vfs-cache-max-age" "60s" "--log-file" "/var/log/rclone/mount-billing.log" "--allow-non-empty" "-vvvv"]
2023/05/22 08:12:04 DEBUG : rclone: Version "v1.54.0-beta.4980.29b553301.topic+yc+debug_version" starting with parameters ["rclone" "mount" "fastone:fastone-billing-mywmicro-fcce-north1-1" "/fastone-mnt" "--config" "/rclone/conf/rclone.conf" "--cache-dir" "/tmp/rclone/.cache" "--dir-cache-time" "2s" "--poll-interval" "1s" "--vfs-cache-mode" "writes" "--vfs-cache-poll-interval" "1s" "--vfs-cache-max-age" "60s" "--log-file" "/var/log/rclone/mount-billing.log" "--allow-non-empty" "--no-modtime" "-vvvv"]
2023/05/25 09:39:30 DEBUG : rclone: Version "v1.54.0-beta.4980.29b553301.topic+yc+debug_version" starting with parameters ["rclone" "mount" "fastone:fastone-billing-mywmicro-fcce-north1-1" "/fastone-mnt" "--config" "/rclone/conf/rclone.conf" "--cache-dir" "/tmp/rclone/.cache" "--dir-cache-time" "2s" "--poll-interval" "1s" "--vfs-cache-mode" "writes" "--vfs-cache-poll-interval" "1s" "--vfs-cache-max-age" "60s" "--log-file" "/var/log/rclone/mount-billing.log" "--allow-non-empty" "--no-modtime" "--vfs-write-back" "0" "-vvvv"]
2023/07/02 02:29:46 DEBUG : rclone: Version "v1.54.0-beta.4980.29b553301.topic+yc+debug_version" starting with parameters ["rclone" "mount" "fastone:fastone-billing-mywmicro-fcce-north1-1" "/fastone-mnt" "--config" "/rclone/conf/rclone.conf" "--cache-dir" "/tmp/rclone/.cache" "--dir-cache-time" "2s" "--poll-interval" "1s" "--vfs-cache-mode" "writes" "--vfs-cache-poll-interval" "1s" "--vfs-cache-max-age" "60s" "--log-file" "/var/log/rclone/mount-billing.log" "--allow-non-empty" "--no-modtime" "-vvvv"]
```

The problem is not fixed in v1.54 and v1.57 - can you try v1.62.2 or later please?

**Comment by ncw** [comment]
I'm going to assume this problem is fixed and close this.
